runtime/compiler/control/CompilationThread.cpp

```diff
@@ -3455,73 +3455,61 @@ void TR::CompilationInfo::stopCompilationThreads()
-            break;
+   TR_ASSERT_FATAL(getNumCompThreadsActive() == 0, "All threads must be inactive at this point\n");
 
-         case COMPTHREAD_SIGNAL_TERMINATE:
-         case COMPTHREAD_STOPPING:
-         case COMPTHREAD_STOPPED:
-            // weird case; we should not do anything
-            break;
+   purgeMethodQueue(compilationSuspended);
 
-         case COMPTHREAD_UNINITIALIZED:
-            // Compilation thread did not have time to become fully initialized
-            curCompThreadInfoPT->setCompilationThreadState(COMPTHREAD_SIGNAL_TERMINATE);
-            break;
+   // Cycle though all non-diagnostic threads and wait for them to terminate. At this point it is possible that a
+   // compilation thread has crashed and is going through the JitDump process. The reason we skipped terminating the
+   // diagnostic threads in the above loop is because the JitDump logic will activate the diagnostic thread to generate
+   // the JitDump, so the diagnostic thread must not be in a terminated state at that point.
+   for (uint8_t i = 0; i < getNumTotalCompilationThreads(); i++)
+      {
+      TR::CompilationInfoPerThread *curCompThreadInfoPT = _arrayOfCompilationInfoPerThread[i];
 
-         default:
-            TR_ASSERT(false, "No other comp thread state possible here\n");
-         } // end switch
+      if (!curCompThreadInfoPT->isDiagnosticThread())
+         {
+         while (curCompThreadInfoPT->getCompilationThreadState() != COMPTHREAD_STOPPED)
+            {
+            getCompilationMonitor()->notifyAll();
+            waitOnCompMonitor(vmThread);
+            }
+         }
       }
 
-   TR_ASSERT_FATAL(getNumCompThreadsActive() == 0, "All threads must be inactive at this point\n");
+   // Now that all compilation threads have terminated, we are sure none of them have crashed, or they have finished
+   // generating a JitDump. Only at this point can we terminate the diagnostic threads.
+   for (uint8_t i = 0; i < getNumTotalCompilationThreads(); i++)
+      {
+      TR::CompilationInfoPerThread *curCompThreadInfoPT = _arrayOfCompilationInfoPerThread[i];
 
-   purgeMethodQueue(compilationSuspended);
+      if (curCompThreadInfoPT->isDiagnosticThread())
+         stopCompilationThread(curCompThreadInfoPT);
+      }
 
    for (uint8_t i = 0; i < getNumTotalCompilationThreads(); i++)
       {
       TR::CompilationInfoPerThread *curCompThreadInfoPT = _arrayOfCompilationInfoPerThread[i];
-      TR_ASSERT(curCompThreadInfoPT, "a thread's compinfo is missing\n");
 
-      while (curCompThreadInfoPT->getCompilationThreadState() != COMPTHREAD_STOPPED)
+      if (curCompThreadInfoPT->isDiagnosticThread())
          {
-         getCompilationMonitor()->notifyAll();
-         waitOnCompMonitor(vmThread);
+         while (curCompThreadInfoPT->getCompilationThreadState() != COMPTHREAD_STOPPED)
+            {
+            getCompilationMonitor()->notifyAll();
+            waitOnCompMonitor(vmThread);
+            }
```

Comment:
Did you ever figure out whether a crashing comp thread will just return to the VM after the jitdump is generated and abrt, or does it actually manage to get back to the `run` loop to terminate properly?

Reply:
The crashing comp thread will run through the entire process in `runJitDump` and will return to the VM which will subsequently terminate the process.

The VM signal handler is the one which caught the original signal and invoked the dump generation sequence. This happens here:
https://github.com/eclipse/openj9/blob/95ba1bfd866d3a68c23c8f01ffd970c5267ef705/runtime/vm/gphandle.c#L843-L848

At the end of this function once all the dumps are generated we will `exit(-1)` the process:
https://github.com/eclipse/openj9/blob/95ba1bfd866d3a68c23c8f01ffd970c5267ef705/runtime/vm/gphandle.c#L891